base/linalg/dense.jl

```diff
@@ -459,12 +459,17 @@ julia> exp(A)
 ```
 """
 exp(A::StridedMatrix{<:BlasFloat}) = exp!(copy(A))
-exp(A::StridedMatrix{<:Integer}) = exp!(float(A))
+exp(A::StridedMatrix{<:Union{Integer,Complex{<:Integer}}}) = exp!(float.(A))
 
 ## Destructive matrix exponential using algorithm from Higham, 2008,
 ## "Functions of Matrices: Theory and Computation", SIAM
 function exp!(A::StridedMatrix{T}) where T<:BlasFloat
     n = checksquare(A)
+    if T <: Real
+        if issymmetric(A)
+            return full(exp(Symmetric(A)))
```

Comment:
Unrelated to this PR; but we should probably implement `exp(::Symmetric)` with the `exp(A) = exp!(copy(A)` pattern, which would allow us to use `exp!` here.